File: projects/dejavu.cs/Dejavu.Language/T1Parser.cs
using System.Text.Json.Nodes;
using Dejavu.Types;

namespace Dejavu.Language;

/// <summary>T1 source → IR JSON document.</summary>
public static class T1Parser
{
    public static JsonObject ParseToIr(string source)
    {
        return new JsonObject
        {
            ["irVersion"] = "1.0",
            ["language"] = IrDefaults.DefaultLanguage.DeepClone(),
            ["body"] = new JsonObject
            {
                ["type"] = "Template",
                ["children"] = new TemplateParser(source).Parse(),
            },
        };
    }

    sealed class TemplateParser(string source, string file = "template.dejavu")
    {
        public JsonArray Parse()
        {
            var (children, _) = ParseBody(0, []);
            return children;
        }

        (JsonArray children, int i) ParseBody(int i, HashSet<string> stop)
        {
            var children = new JsonArray();

            while (i < source.Length)
            {
                if (StartsAt(i, "<%") && stop.Count > 0)
                {
                    var block = ReadCodeBlock(i);
                    var head = CodeHead.Classify(block.Content, block.ContentBase, block.Tokens);
                    if (stop.Contains(head.KindName)) return (children, i);
                }

                if (StartsAt(i, "<#"))
                {
                    var end = FindDelimiter(i + 2, "#>");
                    if (end < 0)
                    {
                        throw new ParseError(
                            "unclosed comment",
                            i,
                            2,
                            file,
                            "comment starts here");
                    }
                    children.Add(new JsonObject
                    {
                        ["type"] = "Comment",
                        ["value"] = source[(i + 2)..end],
                    });
                    i = end + 2;
                    continue;
                }

                if (StartsAt(i, "<%!"))
                {
                    children.Add(new JsonObject
                    {
                        ["type"] = "Text",
                        ["value"] = "<%",
                    });
                    i += 3;
                    continue;
                }

                if (StartsAt(i, "<%"))
                {
                    var open = i;
                    var block = ReadCodeBlock(i);
                    i = block.Next;
                    var head = CodeHead.Classify(block.Content, block.ContentBase, block.Tokens);

                    if (stop.Count > 0 && stop.Contains(head.KindName))
                        return (children, open);

                    switch (head)
                    {
                        case CodeHead.If ifHead:
                        {
                            var (node, ni) = ParseIf(i, ifHead, block.Trim);
                            children.Add(node);
                            i = ni;
                            break;
                        }
                        case CodeHead.Loop loopHead:
                        {
                            var (node, ni) = ParseLoop(i, loopHead, block.Trim);
                            children.Add(node);
                            i = ni;
                            break;
                        }
                        case CodeHead.EndIf or CodeHead.EndLoop or CodeHead.Else or CodeHead.ElseIf:
                            if (stop.Count == 0)
                            {
                                throw new ParseError(
                                    $"unexpected control `{head.KindName}`",
                                    block.ContentBase,
                                    1,
                                    file);
                            }
                            return (children, open);
                        case CodeHead.Expr exprHead:
                            children.Add(new JsonObject
                            {
                                ["type"] = "Interpolation",
                                ["expression"] = ExprParser.Parse(
                                    exprHead.ExprSlice,
                                    source,
                                    file,
                                    exprHead.ExprAbs),
                                ["trim"] = block.Trim,
                            });
                            break;
                    }
                    continue;
                }

                var next = NextMarkup(i);
                if (next is null)
                {
                    children.Add(new JsonObject
                    {
                        ["type"] = "Text",
                        ["value"] = source[i..],
                    });
                    break;
                }
                if (next > i)
                {
                    children.Add(new JsonObject
                    {
                        ["type"] = "Text",
                        ["value"] = source[i..next.Value],
                    });
                    i = next.Value;
                }
                else i++;
            }

            return (children, i);
        }

        (JsonObject node, int i) ParseIf(int i, CodeHead.If head, string trim)
        {
            var test = ExprParser.Parse(head.TestSlice, source, file, head.TestAbs);
            var stop = new HashSet<string> { "else_if", "else", "end_if" };
            var (consequent, pos) = ParseBody(i, stop);
            i = pos;
            var elseIfs = new JsonArray();
            JsonArray? alternate = null;

            while (true)
            {
                var block = ReadCodeBlock(i);
                var h = CodeHead.Classify(block.Content, block.ContentBase, block.Tokens);
                switch (h)
                {
                    case CodeHead.ElseIf elseIf:
                    {
                        i = block.Next;
                        var t = ExprParser.Parse(elseIf.TestSlice, source, file, elseIf.TestAbs);
                        var (body, n) = ParseBody(i, stop);
                        elseIfs.Add(new JsonObject
                        {
                            ["type"] = "Stmt.ElseIf",
                            ["test"] = t,
                            ["consequent"] = body,
                            ["trim"] = "none",
                        });
                        i = n;
                        break;
                    }
                    case CodeHead.Else:
                    {
                        i = block.Next;
                        var (body, n) = ParseBody(i, new HashSet<string> { "end_if" });
                        alternate = body;
                        i = n;
                        var end = ReadCodeBlock(i);
                        if (CodeHead.Classify(end.Content, end.ContentBase, end.Tokens) is not CodeHead.EndIf)
                        {
                            throw new ParseError("expected `end if`", end.ContentBase, 1, file);
                        }
                        i = end.Next;
                        goto Done;
                    }
                    case CodeHead.EndIf:
                        i = block.Next;
                        goto Done;
                    default:
                        throw new ParseError(
                            $"expected if closer, got `{h.KindName}`",
                            block.ContentBase,
                            1,
                            file);
                }
            }

        Done:
            var node = new JsonObject
            {
                ["type"] = "Stmt.If",
                ["test"] = test,
                ["consequent"] = consequent,
                ["elseIfs"] = elseIfs,
                ["trim"] = trim,
            };
            if (alternate is not null) node["alternate"] = alternate;
            return (node, i);
        }

        (JsonObject node, int i) ParseLoop(int i, CodeHead.Loop head, string trim)
        {
            var iterable = ExprParser.Parse(head.IterSlice, source, file, head.IterAbs);
            var (body, pos) = ParseBody(i, new HashSet<string> { "end_loop" });
            i = pos;
            var end = ReadCodeBlock(i);
            if (CodeHead.Classify(end.Content, end.ContentBase, end.Tokens) is not CodeHead.EndLoop)
            {
                throw new ParseError("expected `end loop`", end.ContentBase, 1, file);
            }
            return (new JsonObject
            {
                ["type"] = "Stmt.For",
                ["item"] = head.Item,
                ["iterable"] = iterable,
                ["body"] = body,
                ["trim"] = trim,
            }, end.Next);
        }

        CodeBlock ReadCodeBlock(int i)
        {
            if (!StartsAt(i, "<%"))
                throw new ParseError("expected code open `<%`", i, 1, file);

            var j = i + 2;
            var trim = "none";
            if (j < source.Length)
            {
                var mod = source[j];
                if (mod is '.' or '_' or '-' or '~' or '=')
                {
                    trim = mod switch
                    {
                        '_' => "ws",
                        '-' => "nl",
                        '~' => "ws_nl",
                        '=' => "all",
                        _ => "none",
                    };
                    j++;
                }
            }

            var end = FindDelimiter(j, "%>");
            if (end < 0)
            {
                throw new ParseError("unclosed code block", i, 2, file, "opens here");
            }

            var contentBase = j;
            while (contentBase < end)
            {
                var c = source[contentBase];
                if (c is ' ' or '\t' or '\n' or '\r') contentBase++;
                else break;
            }
            var contentEnd = end;
            while (contentEnd > contentBase)
            {
                var c = source[contentEnd - 1];
                if (c is ' ' or '\t' or '\n' or '\r') contentEnd--;
                else break;
            }

            var content = source[contentBase..contentEnd];
            var tokens = CodeLexer.Lex(content, source, file, contentBase)
                .Where(t => t.Kind != CodeTokenKind.CodeEnd)
                .ToList();
            return new CodeBlock(tokens, content, contentBase, trim, end + 2);
        }

        int FindDelimiter(int from, string delim)
        {
            var d0 = delim[0];
            for (var i = from; i + delim.Length <= source.Length; i++)
            {
                if (source[i] != d0) continue;
                var ok = true;
                for (var k = 1; k < delim.Length; k++)
                {
                    if (source[i + k] != delim[k])
                    {
                        ok = false;
                        break;
                    }
                }
                if (ok) return i;
            }
            return -1;
        }

        int? NextMarkup(int from)
        {
            for (var i = from; i + 1 < source.Length; i++)
            {
                if (source[i] != '<') continue;
                var n = source[i + 1];
                if (n is '%' or '#') return i;
            }
            return null;
        }

        bool StartsAt(int i, string s)
        {
            if (i + s.Length > source.Length) return false;
            for (var k = 0; k < s.Length; k++)
            {
                if (source[i + k] != s[k]) return false;
            }
            return true;
        }

        readonly record struct CodeBlock(
            List<CodeToken> Tokens,
            string Content,
            int ContentBase,
            string Trim,
            int Next);
    }
}

File: projects/dejavu.kt/dejavu-language/src/main/kotlin/dejavu/language/T1Parser.kt
package dejavu.language

import dejavu.types.DEFAULT_LANGUAGE
import kotlinx.serialization.json.*

/** Parse T1 template source into Dejavu IR. */
object T1Parser {
    fun parseToIr(source: String, file: String = "template.dejavu"): JsonObject = buildJsonObject {
        put("irVersion", "1.0")
        put("language", DEFAULT_LANGUAGE)
        put("body", buildJsonObject {
            put("type", "Template")
            put("children", JsonArray(TemplateParser(source, file).parse()))
        })
    }
}

private class TemplateParser(
    private val source: String,
    private val file: String,
) {
    fun parse(): List<JsonElement> {
        val (children, _) = parseBody(0, emptySet())
        return children
    }

    private fun parseBody(i0: Int, stop: Set<CodeHead.Kind>): Pair<List<JsonElement>, Int> {
        val children = mutableListOf<JsonElement>()
        var i = i0

        while (i < source.length) {
            if (source.startsWith("<%", i) && stop.isNotEmpty()) {
                val block = readCodeBlock(i)
                val head = classifyCode(block.content, block.contentBase, block.tokens)
                if (head.kind in stop) return children to i
            }

            if (source.startsWith("<#", i)) {
                val end = findDelimiter(i + 2, "#>")
                if (end < 0) {
                    throw ParseError(
                        message = "unclosed comment",
                        file = file,
                        start = i,
                        length = 2,
                        label = "comment starts here",
                    )
                }
                children += buildJsonObject {
                    put("type", "Comment")
                    put("value", source.substring(i + 2, end))
                }
                i = end + 2
                continue
            }

            if (source.startsWith("<%!", i)) {
                children += buildJsonObject {
                    put("type", "Text")
                    put("value", "<%")
                }
                i += 3
                continue
            }

            if (source.startsWith("<%", i)) {
                val open = i
                val block = readCodeBlock(i)
                i = block.next
                val head = classifyCode(block.content, block.contentBase, block.tokens)

                if (stop.isNotEmpty() && head.kind in stop) {
                    return children to open
                }

                when (head) {
                    is CodeHead.If -> {
                        val (node, ni) = parseIf(i, head, block.trim)
                        children += node
                        i = ni
                    }

                    is CodeHead.Loop -> {
                        val (node, ni) = parseLoop(i, head, block.trim)
                        children += node
                        i = ni
                    }

                    CodeHead.EndIf, CodeHead.EndLoop, CodeHead.Else, is CodeHead.ElseIf -> {
                        if (stop.isEmpty()) {
                            throw ParseError(
                                message = "unexpected control `${head.kind}`",
                                file = file,
                                start = block.contentBase,
                                length = 1,
                            )
                        }
                        return children to open
                    }

                    is CodeHead.Expr -> {
                        children += buildJsonObject {
                            put("type", "Interpolation")
                            put(
                                "expression",
                                parseExpr(head.exprSlice, source = source, file = file, base = head.exprAbs),
                            )
                            put("trim", block.trim)
                        }
                    }
                }
                continue
            }

            val next = nextMarkup(i)
            if (next == null) {
                children += buildJsonObject {
                    put("type", "Text")
                    put("value", source.substring(i))
                }
                break
            }
            if (next > i) {
                children += buildJsonObject {
                    put("type", "Text")
                    put("value", source.substring(i, next))
                }
                i = next
            } else {
                i += 1
            }
        }

        return children to i
    }

    private fun parseIf(i0: Int, head: CodeHead.If, trim: String): Pair<JsonObject, Int> {
        var i = i0
        val test = parseExpr(head.testSlice, source = source, file = file, base = head.testAbs)
        var (consequent, pos) = parseBody(i, setOf(CodeHead.Kind.ElseIf, CodeHead.Kind.Else, CodeHead.Kind.EndIf))
        i = pos
        val elseIfs = mutableListOf<JsonElement>()
        var alternate: List<JsonElement>? = null

        while (true) {
            val block = readCodeBlock(i)
            when (val h = classifyCode(block.content, block.contentBase, block.tokens)) {
                is CodeHead.ElseIf -> {
                    i = block.next
                    val t = parseExpr(h.testSlice, source = source, file = file, base = h.testAbs)
                    val (body, n) = parseBody(i, setOf(CodeHead.Kind.ElseIf, CodeHead.Kind.Else, CodeHead.Kind.EndIf))
                    elseIfs += buildJsonObject {
                        put("type", "Stmt.ElseIf")
                        put("test", t)
                        put("consequent", JsonArray(body))
                        put("trim", "none")
                    }
                    i = n
                }

                CodeHead.Else -> {
                    i = block.next
                    val (body, n) = parseBody(i, setOf(CodeHead.Kind.EndIf))
                    alternate = body
                    i = n
                    val end = readCodeBlock(i)
                    if (classifyCode(end.content, end.contentBase, end.tokens) !is CodeHead.EndIf) {
                        throw ParseError(
                            message = "expected `end if`",
                            file = file,
                            start = end.contentBase,
                            length = 1,
                        )
                    }
                    i = end.next
                    break
                }

                CodeHead.EndIf -> {
                    i = block.next
                    break
                }

                else -> throw ParseError(
                    message = "expected if closer, got `${h.kind}`",
                    file = file,
                    start = block.contentBase,
                    length = 1,
                )
            }
        }

        return buildJsonObject {
            put("type", "Stmt.If")
            put("test", test)
            put("consequent", JsonArray(consequent))
            put("elseIfs", JsonArray(elseIfs))
            if (alternate != null) put("alternate", JsonArray(alternate))
            put("trim", trim)
        } to i
    }

    private fun parseLoop(i0: Int, head: CodeHead.Loop, trim: String): Pair<JsonObject, Int> {
        var i = i0
        val iterable = parseExpr(head.iterSlice, source = source, file = file, base = head.iterAbs)
        val (body, pos) = parseBody(i, setOf(CodeHead.Kind.EndLoop))
        i = pos
        val end = readCodeBlock(i)
        if (classifyCode(end.content, end.contentBase, end.tokens) !is CodeHead.EndLoop) {
            throw ParseError(
                message = "expected `end loop`",
                file = file,
                start = end.contentBase,
                length = 1,
            )
        }
        return buildJsonObject {
            put("type", "Stmt.For")
            put("item", head.item)
            put("iterable", iterable)
            put("body", JsonArray(body))
            put("trim", trim)
        } to end.next
    }

    private data class CodeBlock(
        val tokens: List<CodeToken>,
        val content: String,
        val contentBase: Int,
        val trim: String,
        val next: Int,
    )

    private fun readCodeBlock(i: Int): CodeBlock {
        if (!source.startsWith("<%", i)) {
            throw ParseError(
                message = "expected code open `<%`",
                file = file,
                start = i,
                length = 1,
            )
        }
        var j = i + 2
        var trim = "none"
        val mod = source.getOrNull(j)
        if (mod == '.' || mod == '_' || mod == '-' || mod == '~' || mod == '=') {
            trim = when (mod) {
                '_' -> "ws"
                '-' -> "nl"
                '~' -> "ws_nl"
                '=' -> "all"
                else -> "none"
            }
            j++
        }
        val end = findDelimiter(j, "%>")
        if (end < 0) {
            throw ParseError(
                message = "unclosed code block",
                file = file,
                start = i,
                length = 2,
                label = "opens here",
            )
        }
        var contentBase = j
        while (contentBase < end) {
            val c = source[contentBase].code
            if (c == 0x20 || c == 0x09 || c == 0x0a || c == 0x0d) contentBase++
            else break
        }
        var contentEnd = end
        while (contentEnd > contentBase) {
            val c = source[contentEnd - 1].code
            if (c == 0x20 || c == 0x09 || c == 0x0a || c == 0x0d) contentEnd--
            else break
        }
        val content = source.substring(contentBase, contentEnd)
        val tokens = CodeLexer.lexCode(content, source = source, file = file, base = contentBase)
            .filter { it.kind != CodeTokenKind.CodeEnd }
        return CodeBlock(tokens = tokens, content = content, contentBase = contentBase, trim = trim, next = end + 2)
    }

    /** Scan for a delimiter without using regex. */
    private fun findDelimiter(from: Int, delim: String): Int {
        val d0 = delim[0].code
        var i = from
        while (i + delim.length <= source.length) {
            if (source[i].code != d0) {
                i++
                continue
            }
            var ok = true
            for (k in 1 until delim.length) {
                if (source[i + k].code != delim[k].code) {
                    ok = false
                    break
                }
            }
            if (ok) return i
            i++
        }
        return -1
    }

    private fun nextMarkup(from: Int): Int? {
        var i = from
        while (i + 1 < source.length) {
            if (source[i].code != 0x3c /* < */) {
                i++
                continue
            }
            val n = source[i + 1].code
            if (n == 0x25 /* % */ || n == 0x23 /* # */) return i
            i++
        }
        return null
    }
}
